### ISSUE TITLE
Config parser does not complain about overriden parts of configuration

### ISSUE THREAD

**Issue by RichiH** [prometheus/prometheus#1275]
While I readily agree that autogenerated config files should not have this problem, and that YAML works that way, especially in the trial phase, this can confuse users. Warning or, preferably, erroring out would be appreciated.

If you run with a configuration like:

    scrape_configs:
      - job_name: 'prometheus_system'
        target_groups:
        - targets: ['localhost:9100']

    scrape_configs:
      - job_name: 'foo_system'
        target_groups:
        - targets: ['foot:9100']

the former `scrape_configs` is silently discarded. While that's the obvious case, constructs like

    scrape_configs:
      - job_name: 'prometheus_system'
        target_groups:
        - targets: ['localhost:9100']

    rule_files:
      - 'prometheus.rules'
    
    scrape_configs:
      - job_name: 'foo_system'
        target_groups:
        - targets: ['foo:9100']

are silently discarded as well.

So to summarize: Yes, I fat-fingered, but principle of least surprise would be to error out instead of silently carrying on.

**Comment by fabxc** [comment]
Unfortunately, that is an issue of the YAML parser package. Detecting this in a clean way in Prometheus is not really an option.

If no error is thrown, the YAML spec would actually require any subsequent occurrences to be discarded rather then the previous ones (http://yaml.org/spec/1.1/#id932806).

I agree that it's annoying and an error message would be nice to have. But it's not enough of a problem to introduce silly hacks to the Prometheus code base to work around a bug in an upstream package (for which there's not alternative).

**Comment by RichiH** [comment]
Closing as per IRC.

Created https://github.com/go-yaml/yaml/issues/154 instead.
